core/server/master/src/main/java/alluxio/master/backup/BackupWorkerRole.java

```diff
@@ -0,0 +1,378 @@
+    LOG.info("Received backup message: {}", requestMsg.toString());
+    Preconditions.checkState(!mBackupTracker.inProgress(), "Backup in progress");
+
+    // Reset backup tracker.
+    mBackupTracker.reset();
+    // Update current backup status with given backup id.
+    mBackupTracker.update(new BackupStatus(requestMsg.getBackupId(), BackupState.Initiating));
+    mBackupTracker.updateHostname(NetworkAddressUtils.getLocalHostName(
+        (int) ServerConfiguration.global().getMs(PropertyKey.NETWORK_HOST_RESOLUTION_TIMEOUT_MS)));
+
+    // Start sending backup progress to leader.
+    startHeartbeatThread();
+
+    // Cancel timeout task created by suspend message handler.
+    if (!mBackupTimeoutTask.cancel(true)) {
+      mBackupTracker.updateError(new BackupException("Journal has ben resumed due to a time-out"));
+      return;
+    }
+
+    // Spawn a task for advancing journals to target sequences, then taking the backup.
+    mBackupFuture = mExecutorService.submit(() -> {
+      // Mark state as transitioning.
+      mBackupTracker.updateState(BackupState.Transitioning);
+      try {
+        LOG.info(
+            "Initiating catching up of journals to consistent sequences before starting backup. {}",
+            requestMsg.getJournalSequences());
+        CatchupFuture catchupFuture = mJournalSystem.catchup(requestMsg.getJournalSequences());
+        CompletableFuture.runAsync(() -> catchupFuture.waitTermination())
+            .get(BACKUP_ABORT_AFTER_TRANSITION_TIMEOUT_MS, TimeUnit.MILLISECONDS);
+      } catch (Exception e) {
+        mBackupTracker.updateError(new BackupException(
+            String.format("Failed to catch-up journals: %s", e.getMessage()), e));
+        // Don't resume journals if interrupted.
+        // Catch-up future can be interrupted only when master becomes primary during backup.
+        if (!(e instanceof InterruptedException)) {
+          enforceResumeJournals();
+        }
+        return;
+      }
+      LOG.info("Journal transition completed. Taking a backup.");
+      try {
+        mBackupTracker.updateState(BackupState.Running);
+        AlluxioURI backupUri =
+            takeBackup(requestMsg.getBackupRequest(), mBackupTracker.getEntryCounter());
+        mBackupTracker.updateBackupUri(backupUri);
+        mBackupTracker.updateState(BackupState.Completed);
+        // Wait until backup heartbeats are completed.
+        try {
+          mBackupProgressFuture.get();
+        } catch (Exception e) {
+          LOG.warn("Failed to wait for backup heartbeat completion. ", e);
+        }
+      } catch (IOException e) {
+        mBackupTracker.updateError(
+            new BackupException(String.format("Backup failed at worker: %s", e.getMessage()), e));
+      } finally {
+        enforceResumeJournals();
+      }
+    });
+  }
+
+  /**
+   * Resumes the journals. Crashes the process if resume fails.
+   */
+  private void enforceResumeJournals() {
+    try {
+      mJournalSystem.resume();
```

Comment:
Is there a timing issue where some journals could have been resumed and we cancel the task, or will that cause an exception to be thrown here?

Reply:
There is no such window AFAICT.

Double resuming should be avoided by this line below and the fact that, gaining-primacy first stops this role...

`// Cancel timeout task created by suspend message handler.
    if (!mBackupTimeoutTask.cancel(true)) {
      mBackupTracker.updateError(new BackupException("Journal has been resumed due to a time-out"));
      return;
    }`

Reply:
`mBackupTimeoutTask` can return true if it interrupts `mJournalSystem.resume` while it is in progress right?

Reply:
Reverted try-catch condition in `enforceResumeJournal` to original.